lib/Microservice.js

```diff
@@ -0,0 +1,141 @@
+const loopback = require('loopback');
+const boot = require('loopback-boot');
+
+
+/**
+ * Basic Microservice class wrapping a loopback application.
+ *
+ * @type {module.Microservice}
+ */
+
+const Microservice = module.exports = class Microservice {
+
+    constructor(app, bootOptions = {}) {
+        const defaultAppDir = './app';
+        const defaultConfigDir = `${defaultAppDir}/config`;
+
+        this.app = app;
+        this.server = null;
+        this._bootOptions = Object.assign({
+            appRootDir: defaultAppDir,
+            appConfigRootDir: defaultConfigDir,
+            componentRootDir: defaultConfigDir,
+            dsRootDir: defaultConfigDir,
+            env: 'dev',
+            middlewareRootDir: defaultConfigDir,
+            modelsRootDir: defaultConfigDir,
+        }, bootOptions);
+    }
+
+    /**
+   * Stop the webserver of the microservice.
+   *
+   * @returns {Promise.<Microservice>}
+   */
+    stop() {
+
+        if (!this.server || !this.server.listening) return Promise.reject(new Error('Microservice not started yet'));
+
+        const server = this.server;
+        // avoid race conditions
+        this.server = null;
+
+        return new Promise((resolve, reject) => {
+            server.close((err) => {
+                if (err) {
+                    // if unable to stop the server, reassign it.
+                    this.server = server;
+                    return reject(err);
+                }
+                resolve();
+            });
+        });
+
+    }
+
+    /**
+   * Start the webserver of the microservice.
+   *
+   * @returns {Promise.<Microservice>}
+   */
+    start() {
+
+        if (this.server && this.server.listening) return Promise.resolve(this);
+
+        return new Promise((resolve, reject) => {
+            // @note: this message might throw an exception, we should not catch it,
+            //        due to possible issues in the testing
+            this.server = this.app.listen(() => {
+
+                this.app.emit('started');
+                const baseUrl = this.app.get('url').replace(/\/$/, '');
+                // @todo: replace the logger
```

Comment:
shouldn't we annotate hat where the logger gets created since we probably not going to change the code where we use it?